attestation/commandrun/tracing_tetra.go

```diff
@@ -0,0 +1,636 @@
+		pi:       processInfos{processInfos: []*ProcessInfo{}},
+		fi:       fileInfos{fileInfos: make(map[string]*FileInfo)},
+		si:       socketInfos{socketInfos: make(map[string]*SocketInfo)},
+		ctx:      ctx,
+		done:     make(chan bool),
+		policies: []*Policy{},
+		client:   nil,
+		cr:       cr,
+		pid:      pid,
+	}
+
+	tc.policies = append(tc.policies, GetKProbePolicy(uint(pid), cr.tetragonWatchPrefix))
+
+	binary, err := os.Readlink("/proc/" + fmt.Sprintf("%d", pid) + "/exe")
+	if err != nil {
+		return nil, err
+	}
+
+	var argsStr string
+	args, err := os.ReadFile("/proc/" + fmt.Sprintf("%d", pid) + "/cmdline")
+	if err == nil {
+		argsStr = cleanString(string(args))
+		argsSlice := strings.Split(argsStr, " ")
+		//remove first entry which is the binary name
+		argsSlice = argsSlice[1:]
+		argsStr = strings.Join(argsSlice, " ")
+	}
+
+	digest, err := cryptoutil.CalculateDigestSetFromFile(binary, tc.ctx.Hashes())
+	if err != nil {
+		return nil, err
+	}
+
+	firstEvent := &ProcessInfo{
+		Binary:           binary,
+		Args:             argsStr,
+		ProcessID:        pid,
+		ParentPID:        cr.WitnessPID,
+		BinaryDigest:     digest,
+		StartTime:        time.Now().UTC(),
+		StopTime:         time.Time{},
+		UID:              os.Getuid(),
+		Environ:          "",
+		Flags:            "",
+		processEventType: "",
+	}
+
+	tc.pi.processInfos = append(tc.pi.processInfos, firstEvent)
+
+	return &tc, nil
+}
+
+type Policy struct {
+	Kind       string                    `json:"kind"`
+	APIVersion string                    `json:"apiVersion"`
+	Metadata   MetaData                  `json:"metadata"`
+	Spec       cillium.TracingPolicySpec `json:"spec"`
+}
+
+type MetaData struct {
+	Name string `json:"name"`
+}
+
+func NewClient(ctx context.Context, serverAddress string) (tetragon.FineGuidanceSensorsClient, error) {
+	connCtx, connCancel := context.WithTimeout(ctx, 10*time.Second)
+	defer connCancel()
+	conn, err := grpc.DialContext(connCtx, serverAddress, grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithBlock())
+	if err != nil {
+		log.Errorf("Failed to dial tetragon server: %s", err)
+		return nil, err
+	}
+
+	c := tetragon.NewFineGuidanceSensorsClient(conn)
+	return c, nil
+}
+
+func (tc *TraceContext) Stop(cr *CommandRun) error {
+	err := tc.RemoveAllTraces()
+	if err != nil {
+		return err
+	}
+
+	tc.done <- true
+
+	return nil
+}
+
+func (tc *TraceContext) Start() error {
+	c, err := NewClient(tc.ctx.Context(), tc.cr.tetragonAddress)
+	if err != nil {
+		return err
+	}
+
+	tc.client = c
+
+ADD_POLICY:
+	for _, p := range tc.policies {
+		j, err := json.Marshal(p)
+		if err != nil {
+			return err
+		}
+
+		_, err = tc.client.AddTracingPolicy(tc.ctx.Context(), &tetragon.AddTracingPolicyRequest{
+			Yaml: string(j),
+		})
+		//remove sensor in case it didn't get cleaned up
+		if err != nil {
+			if strings.Contains(err.Error(), "already exists") {
+				RemoveTrace(tc.ctx.Context(), tc.client, p.Metadata.Name)
+				goto ADD_POLICY
```

Comment:
why a goto and label here? can't we just return the err in an if statement else fall through?

Reply:
I was wondering if you were going to notice that.  It is left over after a refactor.  I'll clean it up.